README.md
# IPK project 2 2022/2023 - network sniffer

# Requirements
- g++
- make

# Build and usage
```
make
./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--ndp] [--mld] {-n num}
```

where

- -i eth0 (just one interface to sniff) or --interface. If this parameter is not specified list of active interfaces is printed
- -t or --tcp (will display TCP segments and is optionally complemented by -p functionality).
- -u or --udp (will display UDP datagrams and is optionally complemented by-p functionality).
- -p 23 (extends previous two parameters to filter TCP/UDP based on port number) part of TCP/UDP headers).
- --icmp4 (will display only ICMPv4 packets).
- --icmp6 (will display only ICMPv6 echo request/response).
- --arp (will display only ARP frames).
- --ndp (will display only ICMPv6 NDP packets).
- --igmp (will display only IGMP packets).
- --mld (will display only MLD packets).
- -n 10 (specifies the number of packets to display, i.e., the "time" the program runs; if not specified, consider displaying only one packet, i.e., as if -n 1)

# Theoretical foundations
It is neccessary to understand a few concepts, mainly the packets that are being captured

- TCP (Transmission Control Protocol): a connection-oriented protocol used to establish reliable, ordered, and error-checked streams of data between applications running on hosts on a network. It is a transport layer protocol, and it ensures that data is delivered in the correct order and that no data is lost or corrupted.
- UDP (User Datagram Protocol): a connectionless protocol that operates at the transport layer and is used for sending datagrams over an IP network. Unlike TCP, UDP does not establish a connection before sending data, and it does not provide any error checking or correction mechanisms.
- ARP (Address Resolution Protocol): a protocol used to map an IP address to a physical (MAC) address on a local network. It is used by network devices to send data to other devices on the same network.
- ICMPv4 (Internet Control Message Protocol version 4): a protocol used by network devices to send error messages and operational information about network conditions. It is used for diagnostics and troubleshooting and operates at the network layer of the OSI model.
- ICMPv6 (Internet Control Message Protocol version 6): similar to ICMPv4, but used in IPv6 networks.
- IGMP (Internet Group Management Protocol): a protocol used by hosts and adjacent multicast routers to establish multicast group memberships. It is used to manage multicast group memberships and to control the flow of multicast traffic on a network.
- NDP (Neighbor Discovery Protocol): a protocol used by IPv6 nodes to discover other nodes on the same link, and to determine the link-layer addresses of neighboring nodes. It is used to resolve IPv6 addresses to MAC addresses on a local network.
- MLD (Multicast Listener Discovery): similar to IGMP, but used in IPv6 networks to discover and manage multicast group memberships. It is used to control the flow of multicast traffic in an IPv6 network.

# Program
This program is using mainly pcap.h library for communication with interface and capturing packets. Then a list of other network libraries for parsing packet itself:
- arpa/inet.h
- netinet/tcp.h
- netinet/udp.h
- netinet/ip_icmp.h
- netinet/ether.h
- netinet/ip6.h

## Functions

### Main
``` cpp
int main(int argc, char* argv[])
```
- takes care of program arguments and of program flow

### Signal handler
``` cpp
/**
    * Function for handling Ctrl+C interrupt
    * @param signum - number of signal
*/
void signal_handler(int signum);
```
- reacts when user exits program with Ctrl+C

### Pcap handler
``` cpp
/**
    * Function for creating libpcap handle.
    * @param device - network interface name
    * @param filter - packet filter expression
    * @return pcap_t* - libpcap handle
    * @return NULL - error
    * @source - https://vichargrave.github.io/programming/develop-a-packet-sniffer-with-libpcap/#build-and-run-the-sniffer
*/
pcap_t* create_pcap_handle(char* device, char* filter);
```
- this function opens interface for packet capturing and sets filter (which protocols should be captured)
- it is copied from https://vichargrave.github.io/programming/develop-a-packet-sniffer-with-libpcap/#build-and-run-the-sniffer

### Packet handler
``` cpp
/**
    * Function for handling packets captured by libpcap.
    * @param user - user-defined pointer passed to pcap_loop()
    * @param packethdr - pointer to the packet header struct
    * @param packetptr - pointer to the packet data buffer
*/
void packet_handler(u_char *user, const struct pcap_pkthdr *packethdr, const u_char *packetptr);
```

- handles packet captured by pcap_loop() function and prints packet in correct format

### Create filter

``` cpp
/**
    * Create packet filter expression based on options.
    * @param options - struct containing filter options
    * @return string - packet filter expression
*/
string create_filter(options_t options);

```
- creates filter based on user input


# Testing
Testing was done randomly and probably absolutely unsufficiently, I only started Wireshark with my at the same time sniffer and checked captured packets manually.


# References

[1] Pcap library [online] Available at: https://www.tcpdump.org/pcap.html

[2] Page how to write sniffer [online] Available at: https://vichargrave.github.io/programming/develop-a-packet-sniffer-with-libpcap/#build-and-run-the-sniffer

[3] Macro for argument parsing [online] Available at: https://stackoverflow.com/questions/1052746/getopt-does-not-parse-optional-arguments-to-parameters

[4] ICMPv6 protocol [online] Available at: https://en.wikipedia.org/wiki/ICMPv6

[5] libpcap prezentace FI MUNI [online] Publisher: FI MUNI, Brno, 1. 12. 2015, Jiří Novosad. Available at: https://is.muni.cz/el/fi/podzim2015/PB173/um/59281323/60427338/pb173_linux_10.pdf

and countless pages of stackoverflow questions...

CHANGELOG.md
# Changelog for IPK project 2 - 2023

## Supported functionality
Program is capable of capturin all types of packets according to assignment and handling of Ctrl + C user interrupt signal.

Program arguments can be passed in any order.

## Known limitations
- Program does not support Windows systems

main.cpp
// IPK project 2 - network sniffer
// Author: David Sklenář - xsklen14
// Date: 2023/03/27

#include <getopt.h>
#include <signal.h>
#include <iostream>
#include <vector>
#include <string>
#include <pcap.h>
#include <arpa/inet.h>
#include <netinet/tcp.h>
#include <netinet/ip_icmp.h>
#include <netinet/ether.h>
#include <netinet/ip6.h>

using namespace std;


// stolen from https://stackoverflow.com/questions/1052746/getopt-does-not-parse-optional-arguments-to-parameters
#define OPTIONAL_ARGUMENT_IS_PRESENT \
    ((optarg == NULL && optind < argc && argv[optind][0] != '-') \
     ? (bool) (optarg = argv[optind++]) \
     : (optarg != NULL))


int main(int argc, char* argv[]) {

    static struct option long_options[] = {
        {"interface", optional_argument, 0, 'i'},
        {"tcp", no_argument, 0, 't'},
        {"udp", no_argument, 0, 'u'},
        {"arp", no_argument, 0, 'a'},
        {"icmp4", no_argument, 0, '4'},
        {"icmp6", no_argument, 0, '6'},
        {"igmp", no_argument, 0, 'g'},
        {"mld", no_argument, 0, 'm'},
        {"port", required_argument, 0, 'p'},
        {"num", required_argument, 0, 'n'},
        {0, 0, 0, 0}
    };

    int option_index = 0;
    int opt;

    // Set default values
    string interface_name = "";
    int port = -1;
    int protocol = -1;
    bool arp = false, icmp4 = false, icmp6 = false, igmp = false, mld = false;
    int num = 1;
    while ((opt = getopt_long(argc, argv, "i::p:tua46gmn:", long_options, &option_index)) != -1) {
        //cout << "option: " << static_cast<char>(opt) << endl;
        switch (opt) {
        case 'i':
            if (OPTIONAL_ARGUMENT_IS_PRESENT) {
                interface_name = optarg;
            }
            break;
        case 'p':
            port = atoi(optarg);
            break;
        case 't':
            protocol = IPPROTO_TCP;
            break;
        case 'u':
            protocol = IPPROTO_UDP;
            break;
        case 'a':
            arp = true;
            break;
        case '4':
            icmp4 = true;
            break;
        case '6':
            icmp6 = true;
            break;
        case 'g':
            igmp = true;
            break;
        case 'm':
            mld = true;
            break;
        case 'n':
            num = atoi(optarg);
            break;
        case ':':
            if (optopt == 'i') {
                interface_name = "";
            } else {
                cerr << "Option requires an argument: " << static_cast<char>(optopt) << endl;
            }
            break;
        case '?':
            cerr << "Invalid option: " << static_cast<char>(optopt) << endl;
            break;
        default:
            cerr << "Unknown option: " << static_cast<char>(optopt) << endl;
            break;
        }
    }
    if (interface_name == "" || argc == 1) {
        char error_buffer[PCAP_ERRBUF_SIZE];
        pcap_if_t* all_interfaces;
        if (pcap_findalldevs(&all_interfaces, error_buffer) == -1) {
            cout << "Error: " << error_buffer << endl;
            return 1;
        }
        vector<string> interface_names;
        for (pcap_if_t* interface = all_interfaces; interface; interface = interface->next) {
            interface_names.push_back(interface->name);
        }
        pcap_freealldevs(all_interfaces);
        for (const auto& name : interface_names) {
            cout << name << endl;
        }
        return 0;
    }

    // Print out the parsed arguments
    cout << "interface: " << interface_name << endl;
    cout << "port: " << port << endl;
    cout << "protocol: " << protocol << endl;
    cout << "arp: " << arp << endl;
    cout << "icmp4: " << icmp4 << endl;
    cout << "icmp6: " << icmp6 << endl;
    cout << "igmp: " << igmp << endl;
    cout << "mld: " << mld << endl;
    cout << "num: " << num << endl;

    return 0;
}

ipk-sniffer.h
// IPK project 2 - network sniffer - header file
// Author: David Sklenář - xsklen14
// Date: 2023/03/27

#include <getopt.h>
#include <signal.h>
#include <iostream>
#include <vector>
#include <string>
#include <cstring>
#include <ctime>
#include <pcap.h>
#include <iomanip>
#include <arpa/inet.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/ip_icmp.h>
#include <netinet/ether.h>
#include <netinet/ip6.h>

using namespace std;

// SOURCE: https://stackoverflow.com/questions/1052746/getopt-does-not-parse-optional-arguments-to-parameters
// AUTHOR: https://stackoverflow.com/users/12979602/larsewi
#define OPTIONAL_ARGUMENT_IS_PRESENT \
    ((optarg == NULL && optind < argc && argv[optind][0] != '-') \
     ? (bool) (optarg = argv[optind++]) \
     : (optarg != NULL))

#define MAC_ADDR_STR(mac_addr, str) \
    sprintf(str, "%02x:%02x:%02x:%02x:%02x:%02x", \
            mac_addr[0], mac_addr[1], mac_addr[2], \
            mac_addr[3], mac_addr[4], mac_addr[5])

#define MAX_PORT 65535
#define BYTES_PER_LINE 16

typedef struct {
    string interface_name, arp, icmp4, icmp6, igmp, mld, tcp, udp, ndp;
    int port;
    int num;
} options_t;

options_t options;
pcap_t* handle;


/**
    * Function for handling Ctrl+C interrupt
    * @param signum - number of signal
*/
void signal_handler(int signum);

/**
    * Function for creating libpcap handle.
    * @param device - network interface name
    * @param filter - packet filter expression
    * @return pcap_t* - libpcap handle
    * @return NULL - error
    * @source - https://vichargrave.github.io/programming/develop-a-packet-sniffer-with-libpcap/#build-and-run-the-sniffer
*/
pcap_t* create_pcap_handle(char* device, char* filter);

/**
    * Function for handling packets captured by libpcap.
    * @param user - user-defined pointer passed to pcap_loop()
    * @param packethdr - pointer to the packet header struct
    * @param packetptr - pointer to the packet data buffer
*/
void packet_handler(u_char *user, const struct pcap_pkthdr *packethdr, const u_char *packetptr);

/**
    * Create packet filter expression based on options.
    * @param options - struct containing filter options
    * @return string - packet filter expression
*/
string create_filter(options_t options);



ipk-sniffer.cpp
// IPK project 2 - network sniffer
// Author: David Sklenář - xsklen14
// Date: 2023/03/27

#include "ipk-sniffer.h"

/**
    * Function for handling Ctrl+C interrupt
*/
void signal_handler(int signum) {
    // Terminate program
    pcap_close(handle);
    exit(signum);
}


/**
    * Function for creating libpcap handle.
*/
pcap_t* create_pcap_handle(char* device, char* filter) {
    char errbuf[PCAP_ERRBUF_SIZE];
    pcap_t *handle = NULL;
    struct bpf_program bpf;
    bpf_u_int32 netmask;
    bpf_u_int32 srcip;

    // Get network device source IP address and netmask
    if (pcap_lookupnet(device, &srcip, &netmask, errbuf) == PCAP_ERROR) {
        cerr << "pcap_lookupnet(): " << errbuf << endl;
        return NULL;
    }

    // Open the device for live capture
    handle = pcap_open_live(device, BUFSIZ, 1, 1000, errbuf);
    if (handle == NULL) {
        cerr << "pcap_open_live(): " << errbuf << endl;
        return NULL;
    }

    // Convert the packet filter epxression into a packet filter
    if (pcap_compile(handle, &bpf, filter, 0, netmask) == PCAP_ERROR) {
        cerr << "pcap_compile(): " << pcap_geterr(handle) << endl;
        return NULL;
    }

    // Bind the packet filter to the libpcap handle
    if (pcap_setfilter(handle, &bpf) == PCAP_ERROR) {
        cerr << "pcap_setfilter(): " << pcap_geterr(handle) << endl;
        return NULL;
    }

    return handle;
}


/**
    * Function for handling packets captured by libpcap.
*/
void packet_handler(u_char *user, const struct pcap_pkthdr *packethdr, const u_char *packetptr) {
    if (user == NULL){} //dummy condition just to get rid of warning "unused variable *user"
    // Convert the packet timestamp to a time_t structure
    time_t timestamp_sec = packethdr->ts.tv_sec;
    struct tm *tm_info = localtime(&timestamp_sec);

    // Print the timestamp
    char timestamp[80];
    snprintf(timestamp, 80, "%04d-%02d-%02dT%02d:%02d:%02d.%03ld%+03ld:%02ld", 
         tm_info->tm_year+1900, tm_info->tm_mon+1, tm_info->tm_mday, 
         tm_info->tm_hour, tm_info->tm_min, tm_info->tm_sec, 
         packethdr->ts.tv_usec / 1000, tm_info->tm_gmtoff / 3600, 
         (abs(tm_info->tm_gmtoff) % 3600) / 60);
    cout << "timestamp: " << timestamp << endl;

    //Print source and destination MAC addresses
    struct ether_header *eth_header = (struct ether_header *) packetptr;
    char src_mac[18];
    char dst_mac[18];
    MAC_ADDR_STR(eth_header->ether_shost, src_mac);
    MAC_ADDR_STR(eth_header->ether_dhost, dst_mac);
    cout << "src MAC: " << src_mac << endl;
    cout << "dst MAC: " << dst_mac << endl;

    // Print the packet length
    cout << dec << "frame length: " << packethdr->len << " bytes" << endl;

    uint16_t eth_type = ntohs(eth_header->ether_type);
    if (eth_type == ETHERTYPE_IP) {
        // IPv4
        struct ip *ip_header = (struct ip *) (packetptr + sizeof(struct ether_header));
        char src_ip[INET_ADDRSTRLEN];
        char dst_ip[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &(ip_header->ip_src), src_ip, INET_ADDRSTRLEN);
        inet_ntop(AF_INET, &(ip_header->ip_dst), dst_ip, INET_ADDRSTRLEN);
        cout << "src IP: " << src_ip << endl;
        cout << "dst IP: " << dst_ip << endl;
        if (ip_header->ip_p == IPPROTO_TCP) {
            // TCP
            struct tcphdr *tcp_header = (struct tcphdr *) (packetptr + sizeof(struct ether_header) + sizeof(struct ip));
            uint16_t src_port = ntohs(tcp_header->th_sport);
            uint16_t dst_port = ntohs(tcp_header->th_dport);
            cout << "src port: " << src_port << endl;
            cout << "dst port: " << dst_port << endl;
        
            
        } else if (ip_header->ip_p == IPPROTO_UDP) {
            // UDP
            struct udphdr *udp_header = (struct udphdr *) (packetptr + sizeof(struct ether_header) + sizeof(struct ip));
            uint16_t src_port = ntohs(udp_header->uh_sport);
            uint16_t dst_port = ntohs(udp_header->uh_dport);
            cout << "src port: " << src_port << endl;
            cout << "dst port: " << dst_port << endl;
        } else if (ip_header->ip_p == IPPROTO_ICMP) {
            // ICMP - nothing to add, because it does not work with ports
        }
    } else if (eth_type == ETHERTYPE_IPV6) {
        // IPv6
        struct ip6_hdr *ip6_header = (struct ip6_hdr *) (packetptr + sizeof(struct ether_header));
        char src_ip6[INET6_ADDRSTRLEN];
        char dst_ip6[INET6_ADDRSTRLEN];
        inet_ntop(AF_INET6, &(ip6_header->ip6_src), src_ip6, INET6_ADDRSTRLEN);
        inet_ntop(AF_INET6, &(ip6_header->ip6_dst), dst_ip6, INET6_ADDRSTRLEN);
        cout << "src IP: " << src_ip6 << endl;
        cout << "dst IP: " << dst_ip6 << endl;
        if (ip6_header->ip6_nxt == IPPROTO_TCP) {
            // TCP
            struct tcphdr *tcp_header = (struct tcphdr *) (packetptr + sizeof(struct ether_header) + sizeof(struct ip));
            uint16_t src_port = ntohs(tcp_header->th_sport);
            uint16_t dst_port = ntohs(tcp_header->th_dport);
            cout << "src port: " << src_port << endl;
            cout << "dst port: " << dst_port << endl;
        } else if (ip6_header->ip6_nxt == IPPROTO_UDP) {
            // UDP
            struct tcphdr *tcp_header = (struct tcphdr *) (packetptr + sizeof(struct ether_header) + sizeof(struct ip));
            uint16_t src_port = ntohs(tcp_header->th_sport);
            uint16_t dst_port = ntohs(tcp_header->th_dport);
            cout << "src port: " << src_port << endl;
            cout << "dst port: " << dst_port << endl;
        } else if (ip6_header->ip6_nxt == IPPROTO_ICMPV6) {
            // ICMPv6 - nothing to add, does not work with ports, ndp and mld should be taken care of in filter for pcap_compile
        }
    } else if (eth_type == ETHERTYPE_ARP) {
        // ARP
        struct ether_arp *arp_header = (struct ether_arp *) (packetptr + sizeof(struct ether_header));
        char src_ip[INET_ADDRSTRLEN];
        char dst_ip[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, &(arp_header->arp_spa), src_ip, INET_ADDRSTRLEN);
        inet_ntop(AF_INET, &(arp_header->arp_tpa), dst_ip, INET_ADDRSTRLEN);
        cout << "src IP: " << src_ip << endl;
        cout << "dst IP: " << dst_ip << endl;

    }
    else {
        cout << "Unknown protocol" << endl;
    }

    // printing data
    cout << endl;
    unsigned int byte_offset = 0;
    while (byte_offset < packethdr->len) {
        int remaining_bytes = packethdr->len - byte_offset;
        int bytes_to_print = remaining_bytes < BYTES_PER_LINE ? remaining_bytes : BYTES_PER_LINE;
        cout << "0x" << hex << setfill('0') << setw(4) << byte_offset << ": ";
        for (int i = 0; i < bytes_to_print; i++) {
            cout << hex << setfill('0') << setw(2) << static_cast<int>(packetptr[byte_offset + i]) << " ";
        }
        for (int i = 0; i < BYTES_PER_LINE - bytes_to_print; i++) {
            cout << "   ";
        }
        cout << " ";
        for (int i = 0; i < bytes_to_print; i++) {
            char c = packetptr[byte_offset + i];
            if (isprint(c)) {
                cout << c;
            } else {
                cout << ".";
            }
        }
        cout << endl;
        byte_offset += bytes_to_print;
    }
    cout << endl;
}


/**
    * Create packet filter expression based on options.
*/
string create_filter(options_t options){
    string filter = "";
    bool first = true;
    if (options.arp != "") {
        if (first) {
            filter += "arp";
            first = false;
        } else {
            filter += " or arp";
        }
    }
    if (options.icmp4 != "") {
        if (first) {
            filter += "icmp";
            first = false;
        } else {
            filter += " or icmp";
        }
    }
    if (options.icmp6 != "") {
        if (first) {
            filter += "(icmp6[0] == 128 or icmp6[0] == 129)";
            first = false;
        } else {
            filter += " or (icmp6[0] == 128 or icmp6[0] == 129)";
        }
    }
    if (options.igmp != "") {
        if (first) {
            filter += "igmp";
            first = false;
        } else {
            filter += " or igmp";
        }
    }
    if (options.mld != "") {
        if (first) {
            filter += "(icmp6[0] == 130 or icmp6[0] == 131 or icmp6[0] == 132 or icmp6[0] == 143)";
            first = false;
        } else {
            filter += " or (icmp6[0] == 130 or icmp6[0] == 131 or icmp6[0] == 132 or icmp6[0] == 143)";
        }
    }
    if (options.tcp != "") {
        if (options.port != -1) {
            if (first) {
                filter += "(tcp port " + to_string(options.port) + ")";
                first = false;
            } else {
                filter += " or (tcp port " + to_string(options.port) + ")";
            }
        } else {
            if (first) {
                filter += "tcp";
                first = false;
            } else {
                filter += " or tcp";
            }
        }
    }
    if (options.udp != "") {
        if (options.port != -1) {
            if (first) {
                filter += "(udp port " + to_string(options.port) + ")";
                first = false;
            } else {
                filter += " or (udp port " + to_string(options.port) + ")";
            }
        } else {
            if (first) {
                filter += "udp";
                first = false;
            } else {
                filter += " or udp";
            }
        }
    }
    if (options.ndp != "") {
        if (first) {
            filter += "(icmp6[0] == 133 or icmp6[0] == 134 or icmp6[0] == 135 or icmp6[0] == 136 or icmp6[0] == 137)";
            first = false;
        } else {
            filter += " or (icmp6[0] == 133 or icmp6[0] == 134 or icmp6[0] == 135 or icmp6[0] == 136 or icmp6[0] == 137)";
        }
    }
    return filter;

}

int main(int argc, char* argv[]) {
    string filter = "";

    static struct option long_options[] = {
        {"interface", optional_argument, 0, 'i'},
        {"tcp", no_argument, 0, 't'},
        {"udp", no_argument, 0, 'u'},
        {"arp", no_argument, 0, 'a'},
        {"icmp4", no_argument, 0, '4'},
        {"icmp6", no_argument, 0, '6'},
        {"igmp", no_argument, 0, 'g'},
        {"mld", no_argument, 0, 'm'},
        {"port", required_argument, 0, 'p'},
        {"num", required_argument, 0, 'n'},
        {"ndp", no_argument, 0, 'd'},
        {0, 0, 0, 0}
    };

    int option_index = 0;
    int opt;

    // Set default values
    options.interface_name = "";
    options.port = -1;
    options.arp = "";
    options.icmp4 = "";
    options.icmp6 = "";
    options.igmp = "";
    options.mld = "";
    options.tcp = "";
    options.udp = "";
    options.num = 1;
    while ((opt = getopt_long(argc, argv, "i::p:tua46gmdn:", long_options, &option_index)) != -1) {
        switch (opt) {
        case 'i':
            if (OPTIONAL_ARGUMENT_IS_PRESENT) {
                options.interface_name = optarg;
            }
            break;
        case 'p':
            if (options.port != -1) {
                cerr << "Error: Port is already set." << endl;
                return 1;
            }
            char *ptr;
            options.port = strtol(optarg, &ptr, 10);
            break;
        case 't':
            if (options.tcp != "") {
                cerr << "Error: TCP is already set." << endl;
                return 1;
            }
            options.tcp = "tcp";
            break;
        case 'u':
            if (options.udp != "") {
                cerr << "Error: UDP is already set." << endl;
                return 1;
            }
            options.udp = "udp";
            break;
        case 'a':
            if (options.arp != "") {
                cerr << "Error: ARP is already set." << endl;
                return 1;
            }
            options.arp = "arp";
            break;
        case '4':
            if (options.icmp4 != "") {
                cerr << "Error: ICMP4 is already set." << endl;
                return 1;
            }
            options.icmp4 = "icmp4";
            break;
        case '6':
            if (options.icmp6 != "") {
                cerr << "Error: ICMP6 is already set." << endl;
                return 1;
            }
            options.icmp6 = "icmp6";
            break;
        case 'g':
            if (options.igmp != "") {
                cerr << "Error: IGMP is already set." << endl;
                return 1;
            }
            options.igmp = "igmp";
            break;
        case 'm':
            if (options.mld != "") {
                cerr << "Error: MLD is already set." << endl;
                return 1;
            }
            options.mld = "mld";
            break;
        case 'd':
            if (options.ndp != "") {
                cerr << "Error: NDP is already set." << endl;
                return 1;
            }
            options.ndp = "ndp";
            break;
        case 'n':
            if (options.num != 1) {
                cerr << "Error: Number of packets is already set." << endl;
                return 1;
            }
            options.num = atoi(optarg);
            break;
        case ':':
            cerr << "Option requires an argument: " << static_cast<char>(optopt) << endl;
            break;
        case '?':
            cerr << "Invalid option: " << static_cast<char>(optopt) << endl;
            break;
        default:
            cerr << "Unknown option: " << static_cast<char>(optopt) << endl;
            break;
        }

    }

    // if something is wrong with user input print all available interfaces
    if (options.interface_name == "") {
        char error_buffer[PCAP_ERRBUF_SIZE];
        pcap_if_t* all_interfaces;
        if (pcap_findalldevs(&all_interfaces, error_buffer) == -1) {
            cerr << "Error: " << error_buffer << endl;
            return 1;
        }
        vector<string> interface_names;
        for (pcap_if_t* interface = all_interfaces; interface; interface = interface->next) {
            interface_names.push_back(interface->name);
        }
        pcap_freealldevs(all_interfaces);
        for (const auto& name : interface_names) {
            cerr << name << endl;
        }
        return 0;
    }

    signal(SIGINT, signal_handler);

    // port correct format handling
    if (options.port != -1 && (options.port < 0 || options.port > MAX_PORT)) {
        cerr << "Invalid port number: " << options.port << endl;
        return 1;
    }
    if (options.port != -1 && (options.tcp == "" && options.udp == "")) {
        cerr << "Port number is set but no protocol is set." << endl;
        return 1;
    }

    if (options.num < 0 && options.num != -1) {
        cerr << "Invalid number of packets: " << options.num << endl;
        return 1;
    }

    // apply default filter if no argument is given
    if (options.tcp == "" && options.udp == "" && options.arp == "" && 
        options.icmp4 == "" && options.icmp6 == "" && options.igmp == "" && 
        options.mld == "" && options.ndp == "") {
        filter = "tcp or udp or arp or icmp or (icmp6[0] == 128 or icmp6[0] == 129) or igmp or (icmp6[0] == 130 or icmp6[0] == 131 or icmp6[0] == 132 or icmp6[0] == 143) or (icmp6[0] == 133 or icmp6[0] == 134 or icmp6[0] == 135 or icmp6[0] == 136 or icmp6[0] == 137)";
    }
    else {
        filter = create_filter(options);
    }
    

    // open interface for listening
    handle = create_pcap_handle((char *)options.interface_name.c_str(), (char *)filter.c_str());
    if(handle == NULL) {
        return 1;
    }

    //check used datalink protocol
    int link_type;
    if ((link_type = pcap_datalink(handle)) < 0) {
        cerr << "pcap_datalink():" << pcap_geterr(handle) << endl;
        return -1;
    }
    if (link_type != DLT_EN10MB) {
        cerr << "This program only supports Ethernet." << endl;
        return -1;
    }

    // packet capturing loop
    if (pcap_loop(handle, options.num, packet_handler, (u_char*)NULL) < 0) {
        cerr << "pcap_loop():" << pcap_geterr(handle) << endl;
	    return -1;
    }

    pcap_close(handle);
    return 0;
}
